src/main/java/org/opensearch/plugin/insights/core/reader/LocalIndexReader.java

```diff
@@ -110,69 +116,119 @@ public LocalIndexReader setIndexPattern(DateTimeFormatter indexPattern) {
      * @param id         query/group id
      * @param verbose    whether to return full output
      * @param metricType metric type to read
-     * @return list of SearchQueryRecords whose timestamps fall between from and to
      */
     @Override
-    public List<SearchQueryRecord> read(
+    public void read(
```

Comment:
As a start, I think we need to refactor this function as it is pretty unreadable. This is doing too many things ie cluster state updates, query building, search etc, which in itself should sit inside their own functions or respective classes.

For example, query builder operation can be inside some helper/util class where it accepts indices, start, end etc, and accordingly returns a query builder for topN. And I hope it makes it reusable elsewhere as well in the future.